contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSchemaFactory.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.store.jdbc;
+
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.drill.exec.store.SchemaConfig;
+import org.apache.drill.exec.store.jdbc.clickhouse.ClickhouseCatalogSchema;
+
+public class JdbcSchemaFactory {
+  private final JdbcStoragePlugin plugin;
+
+  public JdbcSchemaFactory (JdbcStoragePlugin plugin) {
+    this.plugin = plugin;
+  }
+
+  public void registerSchemas(SchemaConfig config, SchemaPlus parent) {
+    if (plugin.getConfig().isClickhouse()) {
```

Comment:
Here we've added an `isClickhouse()` method to the base JDBC plugin. We might add five more "isDruid()", "isAthena()", "isNextBigThing()", ... Then we add logic in various places that use those queries.

In doing so, we couple all these dialects and make the code harder to maintain.

Suggestion: add a `getDialect()` method to the config. It returns a `JdbcDialect` subclass. `getDialect()` does what is needed to identify Clickhouse (or whatever) and return the specific dialect. Else, it returns a `GenericDialect` instance.

One of the methods on `JdbcDialect` is `registerSchemas()`. The "then" part of this code goes in the Clickhouse version, the "else" part of this code goes into the `GenericDialect` version.

Voila! No more tight coupling of dialects.

Of course, if there is a better way to solve the problem, feel free to use that instead; the above is a quick & dirty example.

Reply:
@paul-rogers 
Alternatively, the dialect is actually stored in the Plugin file, not the config file.   More good news is that there already is a method in the Plugin file:

https://github.com/apache/drill/blob/39b565f112122734c080324fdcbef518ced16507/contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcStoragePlugin.java#L81-L83

If I'm tracking the logic correctly, you could use this logic to get the dialect without having to modify the config at all.

Reply:
@cgivre I think @paul-rogers means a new "dialect" interface to be implemented to handle the differrent behaviors for databases connected using JdbcStoragePlugin, as registerSchemas and generateSql.